### ISSUE TITLE
Make QAS in NegotiationCard have an explicit link to QAS in backend

### ISSUE THREAD

**Issue by sei-aderr** [mlte-team/mlte#552]
Currently the QASs in the NegotiationCard are entities on their own, not associated to anything else in the system. Moving forward, they will be the linking point between NegotiationCards and TestCases, and also between different version of the model or different TestResults of a model/version.

Modify the QASs to become this type of entity. More specifically, QASs should:

1. Each QAS should have a name/id so it can be referenced from the TestSpec
1. Each QAS object should have one (or more?) Quality Attributes (QAs) associated to it, as a way of tagging it. 
2. QAs themselves should have a QA Category object associated to them (see #588 )

After some discussion (see below) a new type of Store, CustomLists, will be created to generically store the list of QA Categories and QAs. This needs to be done before we can update QAS to be associated to QAs. This will be done in #589.

**Comment by sei-aderr** [comment]
@sei-glewis 

Two related questions,

1. In the test catalog, we have two fields Quality Attribute Category, which is a predefined list with things like robustness, safety etc, and then Quality Attribute which is free form user input. Then in the negotiation card under system requirements, we have the System Quality field. Is system quality intended to be a Quality Attribute Category, or a more specific Quality Attribute?

2. Is it intended for the Quality Attributes and the Quality Attribute Categories to both be predefined lists with the ability for the user to add onto them, free form, or strictly predefined?

**Comment by sei-glewis** [comment]
1. System Quality is a specific quality attribute. I will take a look at the Negotiation Card to make sure that we align.

2. What I would like is for it to be pre-defined lists that users can add to — that makes it easier to search the catalog.

**Comment by sei-aderr** [comment]
@sebastian-echeverria 

Do you have thoughts on the implementation of the linking between frontend and backend? The frontend to backend link should be simple enough, but handling it with the library I am not sure about.

Grace would like it to be a dynamic, so I think we would have to store that data in whatever store implementation that the user is using. I am thinking each would be 2 values, the display value, just the text `Security` or whatever and then an id that would be generated in the backend. 

So then there would have to be endpoints for CRUD on the lists, probably just create, read. Backend would serve the 2 values, frontend shows them and passes back the (id?) and then stores that. Although storing it as an ID in the actual negotiation card/test catalog entry does require a lookup to that storage location each time that we want to present that data.

Now that would be all well and good, but then comes the issue of using them with the TestSpec and the MLTE library. 

If we want to keep the QA Categories as classes I am really not sure how this would all work. The autofill while coding feature of using the classes with the library is very nice, but at the same time having to define the yourself when you want to add a new one is not ideal. The problem with going away from them though, is if you add new QA Categories to the DB from the frontend, it is not able to be found in the library what is available as new options. This does also apply to the class version, but I guess you would just define a class for it? Could potentially be a code snippet provided by the frontend. 

I am pretty on the fence about what I think we would be better. I think no classes would be a very large change to the codebase with the connection not being super clear, but requiring classes still again requires the user to know about that somehow. 

What are your thoughts?

**Comment by sebastian-echeverria** [comment]
I am guessing whatever we do here would also end up applying to  #475.  So whatever we do, it should be generic enough to apply to all lists where we want to do this, or close enough.

So yes, if we want things like the QA Category and QA to be dynamic lists, then they would have to be stored in one of the stores (we'd have to check exactly which one, or maybe a new one with the same interface), since that is the only place where we can store user-added things.  Maybe it could be a specific store of user-defined pairs, associated to different things. And yes, a name and an id could be enough.

Yes, agreed on the endpoints. We would have to think about the delete case. What if someone accidentally mispells something? A workaround may be enough, though. I agree that storing the ID would require an additional lookup, but that could be handled by the backend in some way, without additional trips.

So your next questions ties back to #515. The current method of QAs/Properties being classes make for some hints when writing the Test Specification, but it is the only place where it is used. On the other hand, we want the user to be able to add these things much more flexibly, even if we present a builtin list of initial QA Categories/QAs. If we were to remove the "Property" class and its derived classes, and instead have a table with the names of the QAs (formerly properties), the changes would basically only directly affect the Test Spec (indirectly it would affect the ValidatedSpec and other as well, true). But I see very little value, currently, in having the QAs being classes, since they hold very little information, and we want to be able to show lists of QAs and such for the Test Catalog, which makes having classes which can't be easily listed a pain. I agree that, in the end, it would be a big change, but other than the fact that it would be a big change, I am not sure we have other reasons for not changing it. Can you think of any, other than the autocomplete when building the Spec? We would have to think of how the Spec would look like if the qa/property comes from a list, true, and it may be a bit tricky, but in the end it is probably doable.

Btw, the only other piece of information the Property class is a blurb describing the qa/property. We would need that as well beside the ID and name. But I am not sure how to handle that info for user-defined QAs/properties.

**Comment by sei-aderr** [comment]
Agree that the classes do not hold that much information and provide little functionality except for the autocomplete when building a Spec/TestSpec. That is the main benefit that they provide, my issue is more so I just don't see how we would make it easy for the user to get the qa/property values when they come from a list after removing them as classes.

A big focus of these re-architecture changes was to make the library easier to use and I feel like this would be a change that would potentially make it more difficult, although more flexible. Depends how you weight the previous of having to define your own class vs having to find the list of QA Category/QA and use those strings. I just want to have an idea of how the user would interact with the library if they were no longer classes before we go ahead with the changes. Do you have any ideas how that could work? Do you think it would be simple enough to add an endpoint/function to the library that would provide a list of the QA Category/QA that could then be used to define a Spec/TestSpec? Then also do we see that as a user friendly enough implementation. 

If we also want to have the blurbs to describe the QA Category/QA, we would likely just need it to be a separate form at that point for the user to add them in. It is nice to have those to now what each QA Category/QA is meant to represent, is there a place we had planned for them to be presented to the user? I don't think they are in any way now, and not immediately clear where they would be because the QA Category/QA are presented (only?) in the drop down selects which wouldn't fit descripitons.

One other thing to consider, is going forward how independent from the frontend do we want the library to be if needed? In its current state I would say the library can be used separately as the negotiation card bit is mainly just frontend, although it will be more integrated after the planned tasks on this epic are finished. If we want the library to have the potential to be standalone we would also need a way to add to the QA Categories/QA lists directly from the library if we go that route

**Comment by sei-secheverria** [comment]
One thing I think we didn't discuss was "If we also want to have the blurbs to describe the QA Category/QA, we would likely just need it to be a separate form at that point for the user to add them in...." I agree that they are useful, and since we want to provide as many built-ins as possible, it even makes more sense to keep them. Storing the blurbs alongside the names in a table/store should be simple enough, the question is how to integrate them with the frontend.

One option would be to simply not to allow the user-added QAC/QAs to have this blurb, only for the built ins. Rather, they could have it, but there will be no UI option to do so. This could also be a first step, and later we can figure out how to do that if needed (ie, potentially with a separate CRUD for QAC/QAs, where you can go if you explicitly want to add that blurb).

Also, that blurb could be shown when mousing over the name of the QAC/QA or something, when creating the negotiation card or when used, but how exactly depends on how and where the names will be shown. They should be returned along with the names in the library helper call, though.

Another, slightly related question is will each QA scenario be associated to only one QA Category, and only one QA? Or could they be associated to multiple? From a non-QA point of view, you would think it may be better to add the flexibility to have multiple (though two levels of multiple are confusing). But in practice, I don't think QA scenarios are ever qualified as different QAs... you create two different, similar scenarios for different QAs, so I do not think, at least as my first impression, that this would be needed.

**Comment by sei-glewis** [comment]
I am expecting the piece of the front end where you CRUD categories and QAs to have a field called description. This is going to be useful for the test catalog for users that are not as familiar with QAs.

**Comment by sei-aderr** [comment]
@sei-glewis 

The idea of having them makes sense, it's nice to have the extra info for those users. The debate is how we would want to display them in the frontend. At the moment the only places the QA Category, and the QA are displayed there are in the drop down lists so there isn't a clear place to display their descriptions. We can look into some other options for how we want to do that, but not immediately obvious what would be best.

Do you have strong feelings about Sebastians second question about QA scenarios having multiple QA Category or QA? I agree with Sebastian it would make the most sense for simplicity in the frontend and specifying what scenario is actually being adressed.

**Comment by sei-glewis** [comment]
I was thinking it would be part of the drop down, the QA next to the description.

As far as the second question, it would be one scenario to one category to one QA. However, there can be multiple scenarios per category.

**Comment by sei-aderr** [comment]
From discussion with Sebastian:

- Will make a separate store to keep these lists. They don't fit under any other store and will need to be pre-populated on startup
- This new store will need to be made available to the library similar to artifacts so that MLTE can be used without using the frontend.
- Each entry to each list will be a separate file to take advantage of the FS store pieces that are already implemented
- Base it on the User store as it is the most updated

**Comment by sei-secheverria** [comment]
Closing as all sub-issues were closed
